src/units.cpp

```diff
@@ -464,4 +464,99 @@ double Units::scalingFactor(const UnitsPtr &units1, const UnitsPtr &units2)
     return 0.0;
 }
 
+void createUnitMap(const UnitsPtr &units, std::map<std::string, double> &unitMap, double exp = 1.0)
+{
+    if (units->isBaseUnit()) {
+        auto found = unitMap.find(units->name());
+        if (found == unitMap.end()) {
+            unitMap.emplace(units->name(), 1.0);
+        } else {
+            auto unit = unitMap.find(units->name());
+            unit->second += exp;
+        }
+
+    } else if (isStandardUnitName(units->name())) {
+        auto unit = standardUnitsList.find(units->name());
+        for (const auto &u : unit->second) {
+            if (unitMap.find(u.first) == unitMap.end()) {
+                unitMap.emplace(u.first, u.second);
+            } else {
+                auto ut = unitMap.find(u.first);
+                ut->second += u.second * exp; // adding associated base exponent
+            }
+        }
+
+    } else {
+        for (size_t i = 0; i < units->unitCount(); ++i) {
+            std::string ref;
+            std::string pre;
+            std::string id;
+            double expMult;
+            units->unitAttributes(i, ref, pre, exp, expMult, id);
+            if (isStandardUnitName(ref)) {
+                auto unit = standardUnitsList.find(ref);
+                for (const auto &u : unit->second) {
+                    if (unitMap.find(u.first) == unitMap.end()) {
+                        unitMap.emplace(u.first, u.second);
+                    } else {
+                        auto ut = unitMap.find(u.first);
+                        ut->second += u.second * exp;
+                    }
+                }
+            } else {
+                auto model = owningModel(units);
+                if (model != nullptr) {
+                    auto refUnits = model->units(ref);
+                    if ((refUnits == nullptr) || refUnits->isImport()) {
+                        unitMap.clear();
+                        break;
+                    }
+                    createUnitMap(refUnits, unitMap, exp); // recursively add units to the map
+                }
+            }
+        }
+    }
+}
+
+bool Units::isEquivalentTo(const UnitsPtr &units1, const UnitsPtr &units2)
+{
+    // initial checks
+    if ((units1 == nullptr) || (units2 == nullptr)) {
+        return false;
+    }
+    if ((units1->isImport()) || (units2->isImport())) {
+        return false;
+    }
+
+    std::map<std::string, double> units1Map;
+    createUnitMap(units1, units1Map);
+    std::map<std::string, double> units2Map;
+    createUnitMap(units2, units2Map); // creating the maps to compare units over
```

Comment:
If you are creating something like you are implying with the name `createUnitMap` you should just return the unit map from the function.  So this would look like `UnitsMap units1Map = createUnitMap(units1);` (using the typemap defined above.